src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleRouter.cs

```diff
@@ -0,0 +1,150 @@
+
+                foreach (var endpoint in endpoints)
+                {
+                    var routeValues = new RouteValueDictionary();
+                    var templateMatcher = new TemplateMatcher(TemplateParser.Parse(endpoint.RoutePattern.RawText), new RouteValueDictionary());
+                    if (!templateMatcher.TryMatch(path, routeValues))
+                    {
+                        continue;
+                    }
+
+                    var httpMethodAttribute = endpoint.Metadata.GetMetadata<HttpMethodAttribute>();
+                    if (httpMethodAttribute != null && !httpMethodAttribute.HttpMethods.Any(x => x.Equals(method, StringComparison.OrdinalIgnoreCase)))
+                    {
+                        continue;
+                    }
+
+                    routeCandidates.Add(new KeyValuePair<RouteEndpoint, RouteValueDictionary>(endpoint, routeValues));
+                }
+
+                (RouteEndpoint routeEndpointMatch, RouteValueDictionary routeValuesMatch) = FindRouteEndpoint(context.HttpContext, routeCandidates);
+                if (routeEndpointMatch != null && routeValuesMatch != null)
+                {
+                    if (routeEndpointMatch.RoutePattern?.RequiredValues != null)
+                    {
+                        foreach (var requiredValue in routeEndpointMatch.RoutePattern.RequiredValues)
+                        {
+                            routeValuesMatch.Add(requiredValue.Key, requiredValue.Value);
+                        }
+                    }
+
+                    context.Handler = routeEndpointMatch.RequestDelegate;
+                    context.RouteData = new RouteData(routeValuesMatch);
+                    context.HttpContext.Request.RouteValues = routeValuesMatch;
+                }
+
+                if (context.Handler == null)
+                {
+                    _logger.LogError("Matching route endpoint not found for the given request.");
+                }
+            }
+            catch
+            {
+                throw;
+            }
+        }
+
+        private static (RouteEndpoint routeEndpoint, RouteValueDictionary routeValues) FindRouteEndpoint(
+            HttpContext context,
+            IList<KeyValuePair<RouteEndpoint, RouteValueDictionary>> routeCandidates)
+        {
+            if (routeCandidates.Count == 0)
+            {
+                return (null, null);
+            }
+
+            if (routeCandidates.Count == 1)
+            {
+                return (routeCandidates[0].Key, routeCandidates[0].Value);
+            }
+
+            // Note: When there are more than one route endpoint candidates, we need to find the best match
+            //       by looking at the request method, path, and headers. The logic of finding the best match
+            //       as of now is based on the implementation of FhirController actions and attributes.
+            // TODO: Find a more generic way of implementing the logic.
```

Comment:
Create a user story for this TODO.

Reply:
This is like long term plan, and I guess this will involve redesigning the bundle routing altogether. Created user story - https://microsofthealth.visualstudio.com/Health/_workitems/edit/113807